ndpmon.8
.TH NDPMON 8 "November 2006"
.SH NAME
ndpmon - Neighbor Discovery Protocol Monitor
.SH SYNOPSIS
.na
.B ndpmon
[
.B -i
.I interfacename
] [
.B -f
.I configfile
] [
.B -d
.I dtd_file
] [
.B -F
.I filter
]
.br
.ti +7
[
.B -n
.I number
] [
.B -L
] [
.B -v
] [
.B -h
] [
.B -d
.I dtd_file
]
.br
.ti +7
[
.B -g
.I neighbor_file
]
.ad
.SH DESCRIPTION
.B NDPMon
is  a monitoring software for ipv6 Neighbor Discovery. It syslogs activity and reports by email malicious ND message.
.B NDPMon
uses
.BR libpcap
to listen for icmp6 packets and 
.BR libxml2
to use configuration and neighbor cache files.
.LP
The
.B -i
flag is used to change the default interface eth0.
.LP
The
.B -f
flag is used to change the path of the configuration file.
The default is
.IR /usr/local/ndpmon/config_ndpmon.xml
.LP
The 
.B -e
flag is used to change the path to the DTD file for the configuration file. The default is
.I /usr/local/ndpmon/config_ndpmon.dtd
.LP
The
.B -n
flag uses libpcap to specify a limited number of packet to capture.
.LP
The
.B -F
flag allows to change the default icmp6 filter.
.LP
The
.B -L
flag is used to disable syslog and mail reports. This is used to do a learning phase and constitue the neighbor cache.
.LP
The
.B -v
is used to enable the DEBUG mode.
.LP
The 
.B -d
flag is used to change the path to the DTD file for the neighbor cache. The default is
.I /usr/local/ndpmon/neighbor_list.dtd
.LP
The
.B -g
flag is used to change the path to the neighbor cache. The default is 
.I /usr/local/ndpmon/neighbor_list.xml
.LP
Note that an empty
.I neighbor_cache.xml
file must be created before the first time you run
.BR ndpmon.
.LP 
.BR NDPMon
must be run with
.BR root
rights to work.
.LP
.SH "REPORT MESSAGES"
Here's the list of the report messages generated by
.BR ndpmon:
.TP
.B "wrong couple MAC/IP"
Separately, the MAC and IP addresses are valid, but not as
a couple.
.TP
.B "wrong router mac"
The ethernet address of the RA message is not specified in the 
configuration file.
.TP
.B "wrong router ip"
The ip address of the RA message is not specified in the 
configuration file.
.TP
.B "wrong prefix"
The prefix announced in the RA message is not specified in the 
configuration file.
.TP
.B "wrong router redirect"
The RD message does'nt come from a router specified in the configuration file.
.TP
.B "NA router flag"
The NA specifies a router but isn't one according to the configuration file.
.TP
.B "DAD DOS"
The NA answer to NS to avoid it to get an ip address.
.TP
.B "changed ethernet address"
The host switched to a new ethernet address.
.TP
.B "flip flop"
The ethernet address has changed from the most recently seen address to
the second most recently seen address.
.TP
.B "reused old ethernet address"
The ethernet address has changed from the most recently seen address to
the third (or greater) least recently seen address.
.SH "SYSLOG MESSAGES"
Here are some of the syslog messages;
note that messages that are reported are also sysloged.
.TP
.B "new activity"
This ethernet/ip6 address pair has been announced for last time two
months or more.
.TP
.B "new station"
The ethernet address has not been seen before on the link.
.TP
.B "ethernet broadcast"
The mac ethernet address of the host is a broadcast address.
.TP
.B "ip broadcast"
The ip address of the host is a broadcast address.
.TP
.B "bogon"
The source ip address is not local to the local subnet.
.TP
.B "ethernet mismatch"
The source mac ethernet address didn't match the address announced in
option of the ND message.
.SH FILES
.na
.nh
.nf
config_ndpmon.xml - contains settings which must be fill by the administrator
neighbor_list.xml - neighbor cache: all neighbors known to be on the link 
.ad
.hy
.fi
.SH "SEE ALSO"
.na
.nh
.BR arpwatch (8)
.BR ipv6 (7),
.BR pcap (3),
.BR libxml (3).
.ad
.hy
.SH AUTHOR
Thibault Cholez and Frederic Beck for
MADYNES Project,
Loria,
Fr.
.RE
.SH BUGS
.TP
Please send bug reports to frederic.beck@loria.fr
or thibault.cholez@esial.uhp-nancy.fr 
.LP


CHANGES
 - version 1.4.0 20090813 by Frederic Beck and Thomas Buehring
	* source code passed in fortify
	* code clean ups and reorganization
	* countermeasures plugins (Linux only at the moment)
	* icmplib in the countermeasures plugin
	* doxygen documentation

 - version 1.3d 20080116 by Frederic Beck
 	* Corrected general tests that were not applied only on NDP messages

 - version 1.3b 20070910 by Frederic Beck
 	* Corrected FHS support, especially for Binary distribution
 	* New alerts: wrong RA flags, wrong ipv6 router, wrong link address option...
	* Corrected configure and Makefile to support WEB interface

 - version 1.3a 20070925 by Frederic Beck
 	* create_html_table.py: replaces demopipeprogram.pl and stores the alerts in an XML file
	* discovery_history.dat: each time the neighbor list is updated, a line is added in this file with the time and the number of doscovered nodes. Reinitialized at startup.


 - version 1.3 20070910 by Frederic Beck
 	* NetBSD and MAC OS X support
	* FHS compliance
	* Modified configure and Makefile to respect the FHS
	* Config, cache and definitions files autmatically generated by the configure script

 - version 1.2a 20070828 by Frederic Beck
 	* OpenBSD support
	* 1 problem persists with the mail alrts under OpenBSD (Null body makes the program crash)

 - version 1.2 20070727 by Frederic Beck
	* FreeBSD compatibility (thanks to Janos Mohacsi for the initial patch)
	* Revisited configure and Makefile
	* Possibility to enable or disable -g option when building
	* Timer in pcap_open_live changed from -1 to 1000


 - version 1.1 20070105 by Frederic Beck
 	* Applied patch from Steffen Strueber
	* removed warnings
	* new alerts handling including a pipe functionnality
	* demopipeprogram.pl

 - version 1.0a 20061115 by Frederic Beck
	* Extension of the learning phase with routers' discovery
	* Generation of the according configuration file

 - version 1.0 20061106 by Frederic Beck
	* Modified neighbor and router definition
	* New alerts
	* Code adaptation to this change
	* Modified config and cache files

 - Version 0.1b 20061002 by Frederic Beck
 	* License file added
	* Updated Usage
	* Generic configuration and cache files

 - Version 0.1 updated 2006-09-12 17:08 by Frederic Beck
	* Packaged version with documentation and (un)install targets
	* /etc/init.d/ndpmon script
	* /proc handling

 - Version 0.1 2006-09-11 14:11 by Thibault Cholez
	* Initial Version


ndpmon.sh
#!/bin/sh

### BEGIN INIT INFO
# Provides:          ndpmon
# Required-Start:    $syslog $network
# Required-Stop:     $syslog $network
# Should-Start:      $local_fs
# Should-Stop:       $local_fs
# Default-Start:     
# Default-Stop:      0 1 6
# Short-Description: Launch ndpmon daemon
### END INIT INFO

prefix=/usr/local
exec_prefix=${prefix}
datadir=${prefix}/share
confdir=${prefix}/etc
datadir=${prefix}/share
localstatedir=${prefix}/var

INSTALL_DIR=$prefix/ndpmon
MAN_DIR=/usr/local/share/man/man8
BINARY_DIR=${exec_prefix}/sbin
confprefix=/usr/local/etc
CONF_DIR=$confprefix/ndpmon
dataprefix=/var/local
DATA_DIR=$dataprefix/ndpmon


DAEMON="$BINARY_DIR/ndpmon"
INIT="$CONF_DIR/config_ndpmon.xml"
NAME="ndpmon"
DESC="Neighbor Discovery Protocol Monitor"
HOMEDIR="$INSTALL_DIR"
LOGDIR="/var/log/"
PIDFILE="/var/run/ndpmon.pid"


# Check the existence of required files
test -f $DAEMON || exit 0
test -f $INIT || exit 0

# Identify the distribution
if [ -f /etc/debian_version ]
then
	DISTRIB="DEBIAN"
elif [ -f /etc/fedora-release ]
then
	DISTRIB="FEDORA"
elif [ -f /etc/redhat-release ]
then
	DISTRIB="REDHAT"
fi

if [ $DISTRIB != "DEBIAN" ]
then
	# Source function library.
	. /etc/init.d/functions
fi

start() {
	echo -n $"Starting $DESC"
	if [ $DISTRIB = "DEBIAN" ]
	then
		start-stop-daemon --start --quiet --name $NAME --make-pidfile --pidfile $PIDFILE --exec $DAEMON 1>/dev/null 2>&1 &
	else
		daemon +5 $NAME  1>/dev/null 2>&1 &
	fi
	RETVAL=$?
	sleep 2
	echo
	if [ $RETVAL -ne 0 ]; then
	    return $RETVAL
	fi
}

stop() {	
	echo -n $"Stopping $DESC"
	if [ $DISTRIB = "DEBIAN" ]
	then
		kill `cat $PIDFILE`
	else
		killall -9 $NAME
	fi
	RETVAL=$?
	echo
	if [ $RETVAL -ne 0 ]; then
	    return $RETVAL
	fi
	
}

# See how we were called.
RETVAL=0

case "$1" in
  'start' )
  	start;
	echo
	break;;
  'stop' )
  	stop;
	echo
	break;;
  'restart'  )
  	$0 stop
	$0 start
	echo
	break;;
  'status' )
  	if [ $DISTRIB != "DEBIAN" ]
	then
		status $NAME
		RETVAL=$?
	else
		echo $"Usage: $0 {start|stop|restart}";
	fi
	;;
  *)
	echo $"Usage: $0 {start|stop|status|restart}";
	exit 1;;
esac 

exit $?
